habitat/core/dataset.py

```diff
@@ -236,20 +246,116 @@ def get_splits(
             self.episodes = new_episodes
         return new_datasets
 
-    def sample_episodes(self, num_episodes: int) -> None:
-        """
-        Sample from existing episodes a list of episodes of size num_episodes,
-        and replace self.episodes with the list of sampled episodes.
+
+class EpisodeIterator(Iterator):
+    r"""
+    Episode Iterator class that gives options for how a list of episodes
+    should be iterated. Some of those options are desirable for the internal
+    simulator to get higher performance. More context: simulator suffers
+    overhead when switching between scenes, therefore episodes of the same
+    scene should be loaded consecutively. However, if too many consecutive
+    episodes from same scene are feed into RL model, the model will risk to
+    overfit that scene. Therefore it's better to load same scene consecutively
+    and switch once a number threshold is reached.
+    Currently supports the following features:
+        Cycling: when all episodes are iterated, cycle back to start instead of
+            throwing StopIteration.
+        Cycling with shuffle: when cycling back, shuffle episodes groups
+            grouped by scene.
+        Group by scene: episodes of same scene by grouped and will be loaded
+            consecutively.
+        Set max scene repeat: set a number threshold on how many episodes from
+        the same scene can be loaded consecutively.
+        Sample episodes: sample the specified number of episodes.
+    """
+
+    def __init__(
+        self,
+        episodes: List[T],
+        cycle: bool = True,
+        shuffle_scene_when_cycle: bool = False,
+        group_by_scene: bool = True,
+        max_scene_repeat: int = -1,
```

Comment:
IMO it makes more sense to re-shuffle the list of episodes by scene once every `k` actions taken.  This is logic that would have to live in the `Env` class tho.

Reply:
Maybe `_shuffle_iterator_scene()` should not be an internal helper method then. Per-k-actions shuffling option is not hard to do, though I need to feature out a good way to encode these options and make them configurable.